src/test/java/ProfileTests.java

```diff
@@ -0,0 +1,202 @@
+
+        loginPage.login("sakthibala.sengottiyan+2@testpro.io", "te$t$tudent2");
+
+        //Click Profile
+        homePage.clickProfile();
+
+        //Update Email and Save
+        profilePage.updateName("te$t$tudent2", newProfileName);
+
+        if (homePage.isNotDisplayedSuccessAlert()) {
+
+            Assert.assertEquals(profilePage.getProfileName(), newProfileName);
+        }
+
+    }
+
+        //Change the theme of the app
+    @Test (description="Change Theme in Profile",priority = 5)
+    public void testProfileThemeChange() {
+
+        LoginPage loginPage = new LoginPage(getDriver());
+        HomePage homePage = new HomePage(getDriver());
+        ProfilePage profilePage = new ProfilePage(getDriver());
+
+        loginPage.login(VALID_EMAIL, VALID_PASSWORD);
+
+        //Click Profile
+        homePage.clickProfile();
+
+        profilePage.changeThemeToViolet();
+
+        Assert.assertTrue(homePage.isUpdatedThemeDisplayed(),"Updated theme is displayed");
+
+    }
+
+        //Album cover should display the translucent, blurred overlay
+    @Test (description="Show Blurred Overlay of the Album Art",priority = 6)
+    public void testProfileAlbumArtOverlay() {
+
+        LoginPage loginPage = new LoginPage(getDriver());
+        HomePage homePage = new HomePage(getDriver());
+        ProfilePage profilePage = new ProfilePage(getDriver());
+
+        loginPage.login(VALID_EMAIL, VALID_PASSWORD);
+
+        //Click Profile
+        homePage.clickProfile();
+
+        profilePage.clickShowBlurredAlbumArtOverlay();
+
+        Assert.assertTrue(profilePage.isAlbumArtOverlayDisplayed(),"Blurred overlay of the Album Art is displayed");
+    }
+
+    //'Show Now Playing' notification validation - Functionality is not implemented yet
+
+
+    //'Confirm before closing Koel' alert verification
+    @Test (description="Confirm Closing Koel",priority = 7)
+    public void testConfirmClosingKoel() throws InterruptedException, AWTException {
+
+        LoginPage loginPage = new LoginPage(getDriver());
+        HomePage homePage = new HomePage(getDriver());
+        ProfilePage profilePage = new ProfilePage(getDriver());
+
+        boolean confirmClosingAlertDisplayed = false;
+
+        loginPage.login(VALID_EMAIL, VALID_PASSWORD);
+
+        //Click Profile
+        homePage.clickProfile();
+
+
+        profilePage.clickConfirmClosingKoel();
+
+        try {
+            Robot robot = new Robot();
+            Thread.sleep(2000);
+            // Press keys Ctrl + W
+            robot.keyPress(KeyEvent.VK_CONTROL);
+            robot.keyPress(KeyEvent.VK_W);
+            // Release keys Ctrl + W
+            robot.keyRelease(KeyEvent.VK_CONTROL);
+            robot.keyRelease(KeyEvent.VK_W);
+            Thread.sleep(2000);
+
+            //pressing enter
+            robot.keyPress(KeyEvent.VK_ENTER);
+            //releasing enter
+            robot.keyRelease(KeyEvent.VK_ENTER);
+            getDriver().close();
```

Comment:
Thread. sleep is bad! It blocks the current thread and renders it unusable for further work.
Using Thread. sleep() frequently in an automation framework is not a good practice.

Please use the Explicit Wait command, the WebDriver is directed to wait until a certain condition occurs before proceeding with executing the code.

**Don't use Thread.sleep!** It's not a good way to wait for things to happen, especially in tests. It can make your program act weird. Try other ways, like explicit waiting. It's more reliable for tests.

Here's why you should avoid it:

**1) Unreliable Timing:** Using a fixed sleep duration disregards the actual time needed for an action to complete. System speed, network connectivity, or server load can all affect the execution time, resulting in false negatives or hiding real issues. It's crucial to have tests that accurately reflect the system's behavior.

**2)Reduced Efficiency:** Thread.sleep() adds unnecessary delays to test execution. Even if an action completes before the specified sleep time, the thread remains idle, prolonging the overall test duration. For a large number of waits, this can significantly impact your testing efficiency.

**3) Lack of Synchronization:** Thread.sleep() doesn't consider the actual state of your application or UI elements under test. You may encounter situations where the expected element is not yet available, even after the specified sleep time. This can lead to false results or longer waits than necessary.

By embracing explicit waits, you can enhance the reliability, efficiency, and synchronization of your tests, leading to more robust automation. Say goodbye to hidden bugs and embrace a better testing approach!

PS:  **In real work, Thread.sleep is hardly ever used.** I recommend refactoring all your methods without using Thread.sleep

**HINT**

**Try these methods instead** **Thread.sleep**

**Waits**

```
  public WebElement waitForPresence(By by) {
    return new WebDriverWait(driver, Duration.ofSeconds(5))
        .until(ExpectedConditions.presenceOfElementLocated(by));
  }

  public List<WebElement> waitForPresenceOfAllElements(By by, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.presenceOfAllElementsLocatedBy(by));
  }

  public WebElement waitForPresenceOfNestedElement(By parentLocator, By childLocator, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.presenceOfNestedElementLocatedBy(parentLocator, childLocator));
  }

  public WebElement waitForPresenceOfNestedElement(By parentLocator, By childLocator) {
    return waitForPresenceOfNestedElement(parentLocator, childLocator, 5);
  }

  public List<WebElement> waitForPresenceOfNestedElements(By parentLocator, By childLocator) {
    return new WebDriverWait(driver, Duration.ofSeconds(5))
        .until(ExpectedConditions.presenceOfNestedElementsLocatedBy(parentLocator, childLocator));
  }

  public WebElement waitForVisibility(By by, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.visibilityOfElementLocated(by));
  }

  public List<WebElement> waitUntilAllElementsLocatedByAreVisible(By by) {

    return new WebDriverWait(driver, Duration.ofSeconds(5))
        .until(ExpectedConditions.visibilityOfAllElementsLocatedBy(by));
  }

  public List<WebElement> waitUntilAllElementsLocatedByAreVisible(By by, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.visibilityOfAllElementsLocatedBy(by));
  }

  public boolean waitUntilRequiredElementsAreInvisible(List<WebElement> elements) {
    return new WebDriverWait(driver, Duration.ofSeconds(5))
        .until(ExpectedConditions.invisibilityOfAllElements(elements));
  }

  public boolean waitUntilRequiredElementsAreInvisible(List<WebElement> elements, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.invisibilityOfAllElements(elements));
  }


  public WebElement waitForClickable(By by) {
    return new WebDriverWait(driver, Duration.ofSeconds(3))
        .until(ExpectedConditions.elementToBeClickable(by));
  }

  public WebElement waitForClickable(By by, int timeout) {
    return new WebDriverWait(driver, Duration.ofSeconds(timeout))
        .until(ExpectedConditions.elementToBeClickable(by));
  }


  public void waitNumberOfElementsToBeLessThan(By by, int initialValue) {
    new WebDriverWait(driver, Duration.ofSeconds(5))
        .until(ExpectedConditions
            .numberOfElementsToBeLessThan(by,
                initialValue));
  }
```


**2) ScrollPageDown, ScrollAndClick**


```
public void scrollAndClickJS(By by, int timeout) {
    JavascriptExecutor js = (JavascriptExecutor) driver;
    scrollIntoView(by);
    waitForVisibility(by, timeout);
    js.executeScript("arguments[0].click()", getWrappedWebElement(by));
  }

  public void scrollIntoView(WebElement element) {
    JavascriptExecutor js = (JavascriptExecutor) driver;
    js.executeScript("window.scrollTo(0, 0);");
    Actions actions = new Actions(driver);
    actions.moveToElement(element, element.getSize().getWidth(), element.getSize().getHeight());
    actions.perform();
  }

  public void scrollIntoView(By by) {
    scrollIntoView(getWrappedWebElement(by));
  }

  public WebElement getWrappedWebElement(By by) {
    return waitForPresence(by);
  }
```

**3) Getters (List)**

```
public List<WebElement> getElements(By xpath) {
    List<WebElement> elements = new ArrayList<>();
    List<WebElement> wrappedElements = getWrappedWebElements(xpath);
    int listSize = wrappedElements.size();
    if (listSize == 0) {
      throw new NoSuchElementException("List of requested elements is empty! Try to get the list immediately before use or adjust the locator xpath.");
    }
    for (int i = 1; i <= listSize; i++) {
      WebElement element = driver.findElement(By.xpath(String.format("(%s)[%s]", xpath, i)));
      elements.add(element);
    }
    return elements;
  }

  public List<WebElement> getWrappedWebElements(By elementLocator) {
    return waitForPresenceOfAllElements(elementLocator);
  }

  public List<WebElement> waitForPresenceOfAllElements(By elementLocator) {
    return new WebDriverWait(driver, Duration.ofSeconds(3)).
        until(ExpectedConditions.presenceOfAllElementsLocatedBy(elementLocator));
  }
```
**4) States**

```
public boolean isDisplayed(By by, int time) {
    boolean isDisplayed;
    try {
      isDisplayed = waitForVisibility(by,time) != null;
    } catch (TimeoutException e) {
      isDisplayed = false;
    }
    return isDisplayed;
  }
```


**Use these methods in the BasePage, and then in child classes**

Reply:
Thanks for your feedback!

This 93870 - Confirm Closing Koel test case is about the feature that displays pop-up when closing the browser. As this particular user action of clicking X is outside the scope of HTML tree structure, I don't find a way to automate this using Selenium WebDriver by locating the element and implementing wait. 

As an alternative approach, I have used Java.AWT.Robot to interact with the popup and automate the test case. This is not part of Selenium WebDriver API and hence no option is available to add the WebDriver wait to this robot. 
In general, I don't use Thread. Sleep in the automation. Just to make this particular test case work, I've used Thread. Sleep. 

Please correct me if I'm wrong and advise me of any other option that you think would work in this situation.

Thank you!

Reply:
> Thanks for your feedback!
> 
> This 93870 - Confirm Closing Koel test case is about the feature that displays pop-up when closing the browser. As this particular user action of clicking X is outside the scope of HTML tree structure, I don't find a way to automate this using Selenium WebDriver by locating the element and implementing wait.
> 
> As an alternative approach, I have used Java.AWT.Robot to interact with the popup and automate the test case. This is not part of Selenium WebDriver API and hence no option is available to add the WebDriver wait to this robot. In general, I don't use Thread. Sleep in the automation. Just to make this particular test case work, I've used Thread. Sleep.
> 
> Please correct me if I'm wrong and advise me of any other option that you think would work in this situation.
> 
> Thank you!

Hello! Thank you! Yes, sure!

1) JavaScript Executor: In some cases, you can use JavaScript to close the window, bypassing the need for the Robot class: https://www.lambdatest.com/blog/how-to-use-javascriptexecutor-in-selenium-webdriver/

2) Selenium Actions Class: The Selenium Actions class might offer a more robust way to interact with the browser: https://www.selenium.dev/documentation/webdriver/actions_api/mouse/

If this doesn't work out, then leave it with Robot + Thread.sleep

What a great work!

![Thumbs-up-hand-with-crown-illustration-on-transparent-background-PNG](https://github.com/Tootienahs/A59-QA-Automation-Framework/assets/104761482/20861dce-f8e7-4435-9b92-de6436933d0b)

![Quote](https://github-readme-quotes-bay.vercel.app/quote?theme=dark&layout=socrates)